vllm/worker/cpu_worker.py

```diff
@@ -155,11 +155,51 @@ def __init__(
 
         # Setup OpenMP threads affinity.
         omp_cpuids = envs.VLLM_CPU_OMP_THREADS_BIND
+
         if omp_cpuids == "all":
             self.local_omp_cpuid = "all"
         else:
             self.local_omp_cpuid = omp_cpuids.split("|")[rank]
 
+        # Setup OpenMP thread affinity based on NUMA nodes automatically if no THREAD_BIND setting
+        auto_omp_numa_bind = envs.VLLM_CPU_AUTO_OMP_THREAD_NUMA_BIND
+        if auto_omp_numa_bind and self.local_omp_cpuid == "all":
+            tp_size = self.vllm_config.parallel_config.tensor_parallel_size
+            pp_size = self.vllm_config.parallel_config.pipeline_parallel_size
+            world_size = self.vllm_config.parallel_config.world_size
+
+            from importlib import util
+            libnuma_found = util.find_spec("numa") is not None
+            psutil_found = util.find_spec("psutil") is not None
+            if libnuma_found and psutil_found:
+                from numa import info
+                import psutil
+                cpu_count = psutil.cpu_count(logical=False)
+                cpus_allow_list = psutil.Process().cpu_affinity()
+                numa_size = info.get_num_configured_nodes()
+                cpu_count_per_numa = cpu_count // numa_size
+                num_of_no_bind_cpu = envs.VLLM_CPU_NUM_OF_NO_BIND_CPU if envs.VLLM_CPU_NUM_OF_NO_BIND_CPU < (cpu_count_per_numa//2) else (cpu_count_per_numa//2)
+
+                # check allow node_to_cpus list
+                node_to_cpus= []
+                for i in range(numa_size):
+                    if set(info.node_to_cpus(i)).issubset(cpus_allow_list):
+                        #print(info.node_to_cpus(i))
+                        node_to_cpus.append(info.node_to_cpus(i))
+
+                if world_size > len(node_to_cpus):
+                    logger.info("[ERROR] NO AUTO OMP Bind support because request world size: %d is more than allowed numa_size: %d",
+                            world_size, len(node_to_cpus))
+                else:
+                    rank_to_cpus=str(node_to_cpus[self.rank][0]) + '-' + str(node_to_cpus[self.rank][cpu_count_per_numa - 1 - num_of_no_bind_cpu])
```

Comment:
Building the `rank_to_cpus ` string assumes that CPU ids in `node_to_cpus[self.rank]` are contiguous. But it could be that there are CPUs 0-39,240-279. If so, this means that `rank_to_cpus` would contains CPUs 0-279. The safest option would be to construct the CPU list without assuming anything about how CPUs are numbered inside a NUMA node. (There's a lot variation in different platforms.)

Reply:
@askervin 
We used `info.node_to_cpus(i)` to get the list of CPU ids in a numa node, so we don't assume the CPU id will be continues in a numa node.  

For CPUs 0-39,240-279, we only get node_to_cpus = [0-39]  since 240-279 is also in numa node 0 when HT is on.